### DISCUSSION TITLE
How to visualize a 3D mesh fo traveltime tomography

### DISCUSSION THREAD

**Discussion by sminize** [gimli-org/pyGIMLi#357] [Q&A]
Hello everyone,
i want to adapt the crosshole tomography example (https://www.pygimli.org/_examples_auto/2_seismics/plot_02_crosshole_tomography.html ) using 3D and external data. 
I was able to create the data container. However, I have problems creating the mesh and visualizing it. In the 2D example the mesh is created as follows:
```

refinement = 0.25
x = np.arange(0, bh_spacing + refinement, sensor_spacing * refinement)
y = -np.arange(0.0, bh_length + 3, sensor_spacing * refinement)
mesh = pg.meshtools.createGrid(x, y)

ax, _ = pg.show(mesh, hold=True)
ax.plot(sensors[:, 0], sensors[:, 1], "ro")
```

In order to create a 3D mesh I did (senors has now a third dimension):

```
refinement = 0.25
x = np.arange(0, bh_spacing + refinement, sensor_spacing * refinement)
y = -np.arange(0.0, bh_length + 3, sensor_spacing * refinement)
z = -np.arange(0.0, bh_length + 3, sensor_spacing * refinement)
mesh = pg.meshtools.createGrid(x, y, z)

ax, _ = pg.show(mesh, hold=True)
ax.plot(sensors[:, 0], sensors[:, 1],sensors[:, 1], "ro")
```

However, I get the following error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_77339/1311373777.py in <module>
      8 ax, _ = pg.show(mesh, hold=True)
      9 #ax.show()
---> 10 ax.show(A[:, 0], A[:, 1],A[:, 2], "ro")

~/miniconda3/envs/pg/lib/python3.9/site-packages/pyvista/plotting/plotting.py in show(self, title, window_size, interactive, auto_close, interactive_update, full_screen, screenshot, return_img, use_panel, cpos, height)
   3741             if window_size is None:
   3742                 window_size = self.window_size
-> 3743             self.ren_win.SetSize(window_size[0], window_size[1])
   3744 
   3745         # Render

TypeError: SetSize argument 1: integer argument expected, got float
```

When I try:
`pg.show(mesh)
`
The kernel dies. What could be the reason for this?

Maybe one further question. Is the crosshole tomography example in 3D too expensive to run on a normal desktop PC?

Thank you very much for your help

Best regards,

Simon Schmid

**Comment by halbmy** [comment]
This error is in purely in the pyvista package. Just omit the `pg.show` call and export the mesh (and or results) into a vtk and use ParaView to have a look at it.

**Reply by sminize** [reply]
Thank you very much for your fast reply. What is the syntax for exporting the mesh into vtk?
Something like:
`pg.export(mesh)`?

**Comment by halbmy** [comment]
`mesh.exportVTK("out.vtk")`
Before, you can add any property with
`mesh["resistivity"] = model`
